public/app/features/alerting/unified/utils/rule-id.ts

```diff
@@ -0,0 +1,220 @@
+  };
+}
+
+export function fromRule(ruleSourceName: string, namespace: string, groupName: string, rule: Rule): RuleIdentifier {
+  return {
+    ruleSourceName,
+    namespace,
+    groupName,
+    ruleHash: hashRule(rule),
+  };
+}
+
+export function fromCombinedRule(ruleSourceName: string, rule: CombinedRule): RuleIdentifier {
+  const namespaceName = rule.namespace.name;
+  const groupName = rule.group.name;
+
+  if (rule.rulerRule) {
+    return fromRulerRule(ruleSourceName, namespaceName, groupName, rule.rulerRule);
+  }
+
+  if (rule.promRule) {
+    return fromRule(ruleSourceName, namespaceName, groupName, rule.promRule);
+  }
+
+  throw new Error('Could not create an id for a rule that is missing both `rulerRule` and `promRule`.');
+}
+
+export function fromRuleWithLocation(rule: RuleWithLocation): RuleIdentifier {
+  return fromRulerRule(rule.ruleSourceName, rule.namespace, rule.group.name, rule.rule);
+}
+
+export function equal(a: RuleIdentifier, b: RuleIdentifier) {
+  if (isGrafanaRuleIdentifier(a) && isGrafanaRuleIdentifier(b)) {
+    return a.uid === b.uid;
+  }
+
+  if (isCloudRuleIdentifier(a) && isCloudRuleIdentifier(b)) {
+    return (
+      a.groupName === b.groupName &&
+      a.namespace === b.namespace &&
+      a.rulerRuleHash === b.rulerRuleHash &&
+      a.ruleSourceName === b.ruleSourceName
+    );
+  }
+
+  if (isPrometheusRuleIdentifier(a) && isPrometheusRuleIdentifier(b)) {
+    return (
+      a.groupName === b.groupName &&
+      a.namespace === b.namespace &&
+      a.ruleHash === b.ruleHash &&
+      a.ruleSourceName === b.ruleSourceName
+    );
+  }
+
+  return false;
+}
+
+const cloudRuleIdentifierPrefix = 'cri';
+const prometheusRuleIdentifierPrefix = 'pri';
+
+function escapeDollars(value: string): string {
+  return value.replace(/\$/g, '_DOLLAR_');
+}
+
+function unesacapeDollars(value: string): string {
+  return value.replace(/\_DOLLAR\_/g, '$');
+}
+
+export function parse(value: string, decodeFromUri = false): RuleIdentifier {
+  const source = decodeFromUri ? decodeURIComponent(value) : value;
+  const parts = source.split('$');
+
+  if (parts.length === 1) {
+    return { uid: value };
+  }
+
+  if (parts.length === 5) {
+    const [prefix, ruleSourceName, namespace, groupName, hash] = parts.map(unesacapeDollars);
+
+    if (prefix === cloudRuleIdentifierPrefix) {
+      return { ruleSourceName, namespace, groupName, rulerRuleHash: Number(hash) };
+    }
+
+    if (prefix === prometheusRuleIdentifierPrefix) {
+      return { ruleSourceName, namespace, groupName, ruleHash: Number(hash) };
+    }
+  }
+
+  throw new Error(`Failed to parse rule location: ${value}`);
+}
+
+export function tryParse(value: string | undefined, decodeFromUri = false): RuleIdentifier | undefined {
+  if (!value) {
+    return;
+  }
+
+  try {
+    return parse(value, decodeFromUri);
+  } catch (error) {
+    return;
+  }
+}
+
+export function stringify(identifier: RuleIdentifier, encodeForUri = false): string {
```

Comment:
this is a really odd function :) i'd just remove it and call `encodeURIComponent` where the URL is being formatted to avoid confusion